ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/block/FailedExecutionPool.java

```diff
@@ -0,0 +1,121 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.statetransition.block;
+
+import com.google.common.base.Throwables;
+import java.net.SocketTimeoutException;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.TimeoutException;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import tech.pegasys.teku.infrastructure.async.AsyncRunner;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.exceptions.ExceptionUtil;
+import tech.pegasys.teku.infrastructure.logging.LogFormatter;
+import tech.pegasys.teku.spec.datastructures.blocks.SignedBeaconBlock;
+import tech.pegasys.teku.spec.logic.common.statetransition.results.BlockImportResult;
+import tech.pegasys.teku.storage.server.ShuttingDownException;
+
+public class FailedExecutionPool {
+  private static final Logger LOG = LogManager.getLogger();
+  static final Duration MAX_RETRY_DELAY = Duration.ofSeconds(30);
+  static final Duration SHORT_DELAY = Duration.ofSeconds(2);
+  private final Queue<SignedBeaconBlock> awaitingExecution = new ArrayBlockingQueue<>(10);
+  private final BlockManager blockManager;
+  private final AsyncRunner asyncRunner;
+
+  private Optional<SignedBeaconBlock> retryingBlock = Optional.empty();
+
+  private Duration currentDelay = SHORT_DELAY;
+
+  public FailedExecutionPool(final BlockManager blockManager, final AsyncRunner asyncRunner) {
+    this.blockManager = blockManager;
+    this.asyncRunner = asyncRunner;
+  }
+
+  public synchronized void addFailedBlock(final SignedBeaconBlock block) {
+    if (retryingBlock.isEmpty()) {
+      retryingBlock = Optional.of(block);
+      scheduleNextRetry(block);
+    } else {
+      if (retryingBlock.get().equals(block) || awaitingExecution.contains(block)) {
+        // Already retrying this block.
+        return;
+      }
+      if (!awaitingExecution.offer(block)) {
+        LOG.info(
+            "Discarding block {} as execution retry pool capacity exceeded",
+            LogFormatter.formatBlock(block.getSlot(), block.getRoot()));
+      }
+    }
+  }
+
+  private synchronized void handleExecutionResult(
+      final SignedBeaconBlock block, final BlockImportResult importResult) {
+    if (importResult.hasFailedExecutingExecutionPayload()) {
+      currentDelay = currentDelay.multipliedBy(2);
+      if (currentDelay.compareTo(MAX_RETRY_DELAY) > 0) {
+        currentDelay = MAX_RETRY_DELAY;
+      }
+      if (awaitingExecution.isEmpty() || isTimeout(importResult)) {
+        scheduleNextRetry(block);
```

Comment:
So if we have block with something like resource exhaustion attack or any other kind of failure which looks like timeout, we will continue DoS our EL without any chance to swap payload. Maybe it's better to not handle timeout especially for safety and simplicity

Reply:
I can see the logic there, but that's not really how it works out.  The EL won't stop executing a block when the CL times out and ELs can generally only import one block at a time so if we try to execute a different block it will just be queued behind the first one and also timeout. Whereas if we execute the same block it can be easily deduplicated.  Building up a queue of other blocks to try increases memory usage and makes it even harder for the EL to continue.  Plus by retrying the same block we're most likely to get a cached response once the block finally does finish executing. We should be able to get that cached response even if we wind up executing another block in the mean time but there will be limits to how many import results fit in a cache and how long they're held for.

Reply:
Looking at the Geth code it looks like it will not forget some payload check result even if we query something after (and it sounds reasonable, we rarely make cache for 1 result in development). My point was about case of self-recovery , when client has internal timeout for execution (though Geth code doesn't contain it) or external self-recovery which will kill and restart app without signs of life. In case there exists very bad payload with looong execution we could reproduce EL torture with it again and again. But it will be definitely a network issue at this point, not only for Teku.

Reply:
But the EL can't really time out or it could be rejecting an invalid block, so it has to keep processing until the block is done. We can't be sure if it's still processing or not (the connection may have dropped without sending a proper FIN or something like that) but if the EL is alive it will have to keep processing until it completes the block.  And if it does time out it should remember it as an invalid block (otherwise in PoW it would execute it again when it next sees it on gossip).

So we can't make it any worse, but by retrying the same block we open up some simple optimisations for the EL to avoid building up a list of blocks to execute next.

And as you say, if its feasible to create a block that takes that long to execute then the entire network will be DoSed anyway.